pkg/minikube/update/update.go

```diff
@@ -129,91 +164,68 @@ func getLatestVersionFromGitHub(githubOwner, githubRepo string) (semver.Version,
-	if err != nil {
-		return time.Time{}
-	}
-	timeInFile, err := time.Parse(timeLayout, string(lastUpdateCheckTime))
-	if err != nil {
-		return time.Time{}
-	}
-	return timeInFile
-}
-
-func updateBinary(v semver.Version, downloadBinary, updateLinkPrefix, downloadLinkFormat string) {
-	checksum, err := downloadChecksum(v, downloadBinary, downloadLinkFormat)
-	if err != nil {
-		glog.Errorf("Cannot download checksum: %s", err)
-		os.Exit(1)
-	}
+// downloadAndVerifyArchive downloads the archive of latest version of minishift from GitHub, saves it to a temporary location,
+// and verifies the checksum of downloaded archive with checksum available on GitHub.
+// It returns a string containing path to the downloaded archive.
+// It returns an error if any of the steps failed.
+func downloadAndVerifyArchive(latestVersion semver.Version, archiveName, downloadLinkFormat string) (string, error) {
+	url := fmt.Sprintf(downloadLinkFormat, latestVersion, archiveName)
 
-	currentBinary, err := osext.Executable()
-	if err != nil {
-		glog.Errorf("Cannot find current binary to execute: %s", err)
-		os.Exit(1)
-	}
-
-	url := fmt.Sprintf(downloadLinkFormat, v, downloadBinary)
-	updateBinaryFile(url, checksum)
-
-	env := os.Environ()
-	args := os.Args
-	err = syscall.Exec(currentBinary, args, env)
-	if err != nil {
-		glog.Errorf("Failed to execute updated binary %s: %s", currentBinary, err)
-		os.Exit(1)
-	}
-}
-
-func updateBinaryFile(url string, checksum []byte) {
-	fmt.Println("Downloading updated binary")
+	fmt.Println("Downloading updated archive")
 	httpResp, err := http.Get(url)
 	if err != nil {
-		glog.Errorf("Cannot download binary: %s", err)
-		os.Exit(1)
+		return "", errors.New("Cannot download archive")
 	}
 	defer func() { _ = httpResp.Body.Close() }()
 
-	binary := httpResp.Body
+	updatedArchive := httpResp.Body
 	if httpResp.ContentLength > 0 {
 		bar := pb.New64(httpResp.ContentLength).SetUnits(pb.U_BYTES)
 		bar.Start()
-		binary = bar.NewProxyReader(binary)
+		updatedArchive = bar.NewProxyReader(updatedArchive)
 		defer func() {
 			<-time.After(bar.RefreshRate)
 			fmt.Println()
 		}()
 	}
-	err = update.Apply(binary, update.Options{
-		Hash:     crypto.SHA256,
-		Checksum: checksum,
-	})
+
+	archiveBytes, err := ioutil.ReadAll(updatedArchive)
 	if err != nil {
-		glog.Errorf("Cannot apply binary update: %s", err)
-		err := update.RollbackError(err)
-		if err != nil {
-			glog.Errorf("Failed to rollback update: %s", err)
+		return "", errors.New("Unable to read downloaded archive")
+	}
+
+	// Create a temporary directory inside minishift directory to store archive contents
+	dir, err := ioutil.TempDir("", "download")
+	if err != nil {
+		return "", errors.New("Could not create a temporary directory to store archive contents")
+	}
+
+	var downloadedArchive string
+
+	if runtime.GOOS == "windows" {
+		// Store archive in zip file on Windows
+		downloadedArchive = filepath.Join(dir, "minishift.zip")
```

Comment:
You have the true achieve name in the url, right? In _archiveName_ resp. _url_. Why not getting it from there? This seems to be fragile and makes assumptions which might change.